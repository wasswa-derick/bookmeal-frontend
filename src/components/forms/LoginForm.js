import React from "react";
import validator from "validator";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import FormInput from "./FormInput";

/**
 * @class LoginForm
 * @extends {React.Component}
 */
class LoginForm extends React.Component {
  state = {
    data: {
      email: "",
      password: ""
    },
    errors: {}
  };

  /**
   * @param {Event} evt
   *@returns {null} sets value of target element to the corresponding state value
   */
  onChange = evt => {
    this.setState({
      data: { ...this.state.data, [evt.target.name]: evt.target.value }
    });
  };
  /**
   *@param {Event} evt
   * @returns {null} returns null
   */
  onSubmit = evt => {
    evt.preventDefault();
    const { data } = this.state;
    let errors = this.validateData(data);

    if (Object.keys(errors).length === 0) {
      this.props.handleSubmit(data).catch(err => {
        switch (err.response.status) {
          case 500:
            errors.form = "Internal server error occured";
            break;
          case 400:
            errors = { ...err.response.data.errors };
            break;
          default:
            break;
        }
        this.setState({ errors });
      });
    }

    this.setState({ errors });
  };
  /**
   *@param {Object} data
   * @returns {Object} errors object
   */
  validateData = data => {
    const errors = {};
    if (validator.isEmpty(data.email)) {
      errors.email = "This field is required";
    } else if (!validator.isEmail(data.email)) {
      errors.email = "This email is invalid";
    }
    if (validator.isEmpty(data.password)) {
      errors.password = "This field is required";
    }
    return errors;
  };

  /**
   * @returns {form} for Login
   */
  render() {
    const { data, errors } = this.state;
    return (
      <form autoComplete="off" onSubmit={this.onSubmit}>
        {errors.form && (
          <div className="alert alert-danger" role="alert">
            {errors.form}
          </div>
        )}
        <FormInput
          name="email"
          type="email"
          label="Email"
          value={data.email}
          onChange={this.onChange}
          error={errors.email}
        />
        <FormInput
          name="password"
          type="password"
          label="Password"
          value={data.password}
          onChange={this.onChange}
          error={errors.password}
        />
        <button type="submit" className="btn btn-primary btn-submit">
          Create Account
        </button>
        <p>
          Do not have an account?
          <Link href="/signup" to="/signup">
            Signup
          </Link>
        </p>
      </form>
    );
  }
}

LoginForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

export default LoginForm;
