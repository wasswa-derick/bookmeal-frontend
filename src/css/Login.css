@import url("Common.css");
.title {
    text-align: center;
    padding: 50px 0 20px;
}


/* body {
    background: #f6f6f8;
} */

.card-body h3 {
    color: #00a1e0;
}

.auth-card {
    width: 25rem;
}


/*

.title p {
    font-weight: bold;
} */

.login-panel {
    display: grid;
    justify-content: center;
    align-items: center;
}


/* .login-panel {
    width: auto;
    height: auto;
    border: 1px solid #ced4da;
    background: #fff;
} */


/* .input-container {
    padding: 20px 20px 0px 20px;
}

.container {
    display: grid;
    justify-content: center;
    align-items: center;
} */