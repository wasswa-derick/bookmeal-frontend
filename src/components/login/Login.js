import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { LoginForm } from "../common/forms";
import { loginUser } from "../../actions/auth";
/**
 * @class Login
 * @extends {React.Component}
 */
export class Login extends React.Component {
  /**
   * @param {object} data to submit
   * @returns {Promise} promise to the caller
   * @memberof Login
   */
  handleSubmit = data =>
    this.props.loginUser(data).then(() => this.props.history.push("/"));

  /**
   * @returns {any} renders a page
   * @memberof Login
   */
  render() {
    return (
      <div>
        <div className="container login-panel">
          <div className="card auth-card">
            <div className="card-body">
              <h3 className="card-title title">Log Into Your Account</h3>
              <LoginForm handleSubmit={this.handleSubmit} />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Login.propTypes = {
  loginUser: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};
export default connect(null, { loginUser })(Login);
