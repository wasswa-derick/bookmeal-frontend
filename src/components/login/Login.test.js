import React from "react";
import { shallow } from "enzyme";
import { Login } from "./Login";
import LoginForm from "../../components/common/forms/loginForm/LoginForm";

describe("LoginPage", () => {
  let wrapper;
  let mockFn;
  let history;
  beforeEach(() => {
    mockFn = jest.fn();
    history = {
      push: mockFn
    };
    const fn = () => Promise.resolve();
    wrapper = shallow(<Login loginUser={fn} history={history} />);
  });

  it("should render a login form", () => {
    const form = <LoginForm handleSubmit={wrapper.instance().handleSubmit} />;
    expect(wrapper.containsMatchingElement(form)).toEqual(true);
  });

  it("should handle submit", () => {
    const data = { username: "test", password: "test" };
    wrapper.instance().handleSubmit(data);
    expect(wrapper.instance().props.history).toEqual(history);
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
