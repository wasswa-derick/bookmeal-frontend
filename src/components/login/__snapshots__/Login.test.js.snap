// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should render correctly 1`] = `
<div>
  <div
    className="container login-panel"
  >
    <div
      className="card auth-card"
    >
      <div
        className="card-body"
      >
        <h3
          className="card-title title"
        >
          Log Into Your Account
        </h3>
        <LoginForm
          handleSubmit={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
